### PR TITLE
Mdns: separate query responders for advertisers

### LINKED ISSUE #7175: Fix multiple A/AAAA records with same address on minimal mdns
#### Problem
Minimal mdns sending many redundant A / AAAA records

 #### Proposed Solution
Let's just send one.

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
The current setup uses a single query responder for commissionable
and operational and all the records are cleared together and added
in a large group. This makes it challenging to run commissionable
and operational at the same time since they can't be cleared
individually and because there are duplicate records between
operational and commissionable. 
Fixes #7175

#### Change overview
This commit changes the code to use one query responder per advertisement type and gets the
ResponseSender to loop through the query responders to filter for answers.

#### Testing
- Unit tests added for multiple query responders. Single responder tests still pass.
- Manually tested on M5. avahi, chip-device-ctrl both show correct records, confirmed with wireshark
- when the clear calls are removed, can have operational and commissionable running with no duplicate records.

### CODE REVIEW

**Comment by cecille** [comment]
@chrisdecenzo @sweetymhaiske - This PR follows the Clear() patterns used on master currently, but I think it conflicts with
https://github.com/project-chip/connectedhomeip/pull/7397. 

Chris - I'm not totally clear about what the problem was for TVs - can you elaborate and check this PR to see if this is going to cause you issues?

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 1]
```diff
@@ -195,31 +203,36 @@ CHIP_ERROR AdvertiserMinMdns::Start(chip::Inet::InetLayer * inetLayer, uint16_t
...
+    mQueryResponderAllocatorCommissioner.Clear();
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR AdvertiserMinMdns::Advertise(const OperationalAdvertisingParameters & params)
 {
-    mQueryResponderAllocator.Clear();
+    // TODO: When multi-admin is enabled, commissionable does not need to be cleared here.
+    mQueryResponderAllocatorOperational.Clear();
+    mQueryResponderAllocatorCommissionable.Clear();
```
Comment (chrisdecenzo) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:216]: We do not want to clear these. Commissionable and Commissioner are independent of Operational
```suggestion
```

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 2]
```diff
@@ -195,31 +203,36 @@ CHIP_ERROR AdvertiserMinMdns::Start(chip::Inet::InetLayer * inetLayer, uint16_t
...
+    mQueryResponderAllocatorCommissioner.Clear();
```
Comment (chrisdecenzo) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:217]:
```suggestion
```

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 3]
```diff
@@ -265,30 +279,38 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const OperationalAdvertisingParameters &
 
 CHIP_ERROR AdvertiserMinMdns::Advertise(const CommissionAdvertisingParameters & params)
 {
-    mQueryResponderAllocator.Clear();
+    // TODO: When multi-admin is enabled, operational does not need to be cleared here.
+    mQueryResponderAllocatorOperational.Clear();
```
Comment (chrisdecenzo) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:283]:
```suggestion
```

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 4]
```diff
@@ -265,30 +279,38 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const OperationalAdvertisingParameters &
...
+    mQueryResponderAllocatorCommissionable.Clear();
```
Comment (chrisdecenzo) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:284]:
```suggestion
if (params.GetCommissionAdvertiseMode() == CommssionAdvertiseMode::kCommissionableNode)
{
    mQueryResponderAllocatorCommissionable.Clear();
}
else
{
    mQueryResponderAllocatorCommissioner.Clear();
}
```

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 5]
```diff
@@ -265,30 +279,38 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const OperationalAdvertisingParameters &
...
+    mQueryResponderAllocatorCommissionable.Clear();
```
Comment (chrisdecenzo) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:285]:
```suggestion
```

File: src/lib/mdns/minimal/ResponseSender.h [hunk 6]
```diff
@@ -89,7 +89,10 @@ class ResponseSendingState
 class ResponseSender : public ResponderDelegate
 {
 public:
-    ResponseSender(ServerBase * server, QueryResponderBase * responder) : mServer(server), mResponder(responder) {}
+    static constexpr size_t kMaxQueryResponders = 3;
```
Comment (chrisdecenzo) [inline src/lib/mdns/minimal/ResponseSender.h:92]: Does this correspond to the max number of subtypes or the number of service types?

**Review by chrisdecenzo — CHANGES_REQUESTED** [review]
please see comments on not clearing unrelated service types.

File: src/lib/mdns/minimal/ResponseSender.h (hunk 6 above)
Comment (cecille) [reply src/lib/mdns/minimal/ResponseSender.h:92]: Max number of service types - operational, commissionable node, commssioner. I've hardcoded it here - technically I suppose I could put it as a template, but that makes things a bit messier.

**Comment by woody-apple** [comment]
Real test failures here

File: src/lib/mdns/minimal/ResponseSender.h (hunk 6 above)
Comment (msandstedt) [reply src/lib/mdns/minimal/ResponseSender.h:92]: Isn't there a distinct service advertised for each fabric.  So the number of advertisements is actually at least `n fabrics + 1 + 1`, no?

Comment (cecille) [reply src/lib/mdns/minimal/ResponseSender.h:92]: yup - that's correct. Right now we don't have any multi-admin stuff, but you're right that we should be pro-active for when that rolls in for TE4. I think the real solution here is to template it it somehow, but I'm not sure the SDK in general has a concept of # supported fabrics at the moment. It'll have to eventually though.

Proposal for now - per https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/2551, the number on the table right now for minimum # fabrics is 5. Suggest I update this to be hardcoded 7 with a comment linking back to the spec discussion. I opened an issue here to track updating this to get the real number from the SDK config once that work lands. Issue is here: https://github.com/project-chip/connectedhomeip/issues/8000. Does that work for you?

Comment (msandstedt) [reply src/lib/mdns/minimal/ResponseSender.h:92]: Works for me.

**Comment by woody-apple** [comment]
@saurabhst @msandstedt @mspang @jmartinez-silabs @Damian-Nordic ?

**Review by woody-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Pull request merged**